src/engine/sync/enginesync.cpp

```diff
@@ -25,15 +25,67 @@
 #include "engine/sync/internalclock.h"
 #include "util/assert.h"
 
+const bool SYNC_DEBUG = false;
+
 EngineSync::EngineSync(UserSettingsPointer pConfig)
         : BaseSyncableListener(pConfig) {
 }
 
 EngineSync::~EngineSync() {
 }
 
+Syncable* EngineSync::pickMaster(Syncable* enabling_syncable) {
+    std::vector<Syncable*> stopped_sync_decks;
```

Comment:
This is called from the engine thread. We must not use dynamic allocated memory there. 
It would be probably better not to pick the Master during the engine call. I am thinking to a atomic struct, where this and similar is calculated in the in other thread and is copied to the engine in onCallbackStart() or somehwere.

Reply:
ok I will work on this.

Reply:
This function doesn't get called very often -- only when a deck starts or stops or a track is loaded/unloaded. Furthermore these vectors are tiny -- 2 4-element vectors of pointers.  Precalculating this value on every audio callback would be much, much slower, more than the overhead of these two small vectors getting created. 

Would it be acceptable to preallocate the vectors using reserve, as recommended here: https://www.embedded.com/how-to-make-c-more-real-time-friendly/.  In practice we only ever use 4 values (we only count PrimaryDecks).  The compiler will probably keep these 2*4*4 = 32 bytes set aside for us.

Reply:
Yes, we just to make sure that no potential locking malloc is called from under the hood.

Reply:
ok I also finally defined kMaxNumberOfDecks (not sure if I put it in the best place though).

Reply:
Has this been resolved?

Reply:
I made the vectors static so they'll only be allocated the very first time.

Reply:
Those allocations must be move out of the engine thread, no matter when or how often they occur. According to the docs `reserve()` will allocate the requested capacity.

Using `std::vector` is still discouraged in real-time threads, because it could still allocate at any time if you didn't reserve enough capacity. We would never notice this, other than if users start complaining about rare drop outs. At least wrap all push_back() operations into utility functions and assert that size < capacity!

Reply:
I disagree that our realtime standard is "no allocation of vectors in the engine thread."  That's an overly restrictive rule that has not been backed up by any actual proof of performance harm.  Doing **a lot** of allocation in a realtime thread is dangerous, for instance the RubberBand library does a ton of allocation and causes a lot of CPU slowdown in mixxx because of its overuse of malloc in the critical path.  But to broaden that rule to apply to the creation of 2 4-element vectors of pointers (literally tens of bytes) is overly conservative.  Here's another thread that talks about how to think about this problem: https://stackoverflow.com/questions/2819059/preallocating-memory-with-c-in-realtime-environment

Even so, in this case I made the vectors static, so they are only allocated on the first buffer call.